vault.go

```diff
@@ -8,92 +8,132 @@ import (
-
-var KVSource plugin
-
-func getToken() (string, error) {
-	t, exists := os.LookupEnv("VAULT_TOKEN")
-	if !exists {
-		tokenPath, exists := os.LookupEnv("VAULT_TOKEN_PATH")
-		if exists == false {
-			return "", errors.New("No vault token and vault token path")
-		}
+type VaultSecret struct {
+	Path    string
+	Key     string
+	Relabel string
+}
 
-		tBytes, err := ioutil.ReadFile(tokenPath)
-		if err != nil {
-			fmt.Println("Could not read Vault token from $VAULT_TOKEN_PATH")
-			return "", err
-		}
+type plugin struct {
+	rf               *resmap.Factory
+	ldr              ifc.Loader
+	types.ObjectMeta `json:"metadata,omitempty" yaml:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+	Secrets          []VaultSecret `json:"secrets,omitempty" yaml:"secrets,omitempty"`
+	VaultToken       string
+	VaultAddr        string
+	VaultClient      *api.Client
+}
 
-		t = strings.TrimSpace(string(tBytes))
-		if len(t) == 0 {
-			fmt.Println("Vault token file is empty")
-			return "", err
-		}
-	}
+//nolint: golint
+//noinspection GoUnusedGlobalVariable
+var KustomizePlugin plugin
 
-	return t, nil
+var database = map[string]string{
+	"secret/data/prd/am1/kube0/newman-api": "SaturnV",
 }
 
-func (p plugin) Get(root string, args []string) (map[string]string, error) {
-	var vaultAddr = os.Getenv("VAULT_ADDR")
+func (p *plugin) Config(ldr ifc.Loader, rf *resmap.Factory, c []byte) error {
+	vaultAddr, ok := os.LookupEnv("VAULT_ADDR")
+	if !ok {
+		return errors.New("Missing `VAULT_ADDR` env var: required")
+	}
 
-	var token, err = getToken()
+	vaultToken, err := getToken()
 	if err != nil {
-		panic(err)
+		return err
 	}
 
 	config := &api.Config{
 		Address: vaultAddr,
 	}
 
 	client, err := api.NewClient(config)
-
 	if err != nil {
-		fmt.Println(err)
-		return nil, err
+		return err
 	}
 
-	client.SetToken(token)
+	client.SetToken(vaultToken)
 
-	mapOfSecrets := make(map[string]string)
+	p.rf = rf
+	p.ldr = ldr
+	p.VaultToken = vaultToken
+	p.VaultAddr = vaultAddr
+	p.VaultClient = client
 
-	for _, k := range args {
-		var t = strings.Split(k, "//")
-		var path = t[0]
-		var splitPath = strings.Split(path, "/")
-		var secretPrefix = splitPath[len(splitPath)-1]
-		var key = t[1]
-		var splitKey = strings.Split(key, "@")
+	return yaml.Unmarshal(c, p)
+}
 
-		secret, err := client.Logical().Read(path)
+func (p *plugin) Generate() (resmap.ResMap, error) {
+	args := types.SecretArgs{}
+	args.Name = p.Name
+	args.Namespace = p.Namespace
+
+	for _, secret := range p.Secrets {
+		value, err := p.getSecretFromVault(secret.Path, secret.Key)
 		if err != nil {
-			fmt.Println(err)
 			return nil, err
 		}
-		if secret == nil {
-			return nil, fmt.Errorf("the path %q was not found", path)
-		}
 
-		vaultSecretData, ok := secret.Data["data"].(map[string]interface{})
-		if !ok {
-			return nil, fmt.Errorf("malformed ecret data: %q", secret.Data["data"])
+		var key string
+		if secret.Relabel != "" {
+			key = secret.Relabel
+		} else {
+			key = secret.Key
```

Comment:
Didn't this used to be the `{secret_name}_{secret_key}`?

Reply:
Yes, but now that we have a better API, we have the option to have nicer names in `secretKeyRef` blocks. It requires more changes now because we have to touch a bunch of controller manifests, but it makes `secretKeyRefs` a little simpler too.